File: android/src/main/kotlin/com/hyperpay/sdk/HyperpaySdkPlugin.kt
package com.hyperpay.sdk

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResultLauncher
import com.oppwa.mobile.connect.checkout.dialog.CheckoutActivity
import com.oppwa.mobile.connect.checkout.dialog.CheckoutActivityResult
import com.oppwa.mobile.connect.checkout.dialog.CheckoutActivityResultContract
import com.oppwa.mobile.connect.checkout.meta.CheckoutSettings
import com.oppwa.mobile.connect.checkout.meta.CheckoutSkipCVVMode
import com.oppwa.mobile.connect.exception.PaymentError
import com.oppwa.mobile.connect.exception.PaymentException
import com.oppwa.mobile.connect.payment.BrandsValidation
import com.oppwa.mobile.connect.payment.CheckoutInfo
import com.oppwa.mobile.connect.payment.card.CardPaymentParams
import com.oppwa.mobile.connect.payment.token.TokenPaymentParams
import com.oppwa.mobile.connect.provider.Connect
import com.oppwa.mobile.connect.provider.ITransactionListener
import com.oppwa.mobile.connect.provider.OppPaymentProvider
import com.oppwa.mobile.connect.provider.Transaction
import com.oppwa.mobile.connect.provider.TransactionType
import com.oppwa.mobile.connect.provider.ThreeDSWorkflowListener

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import com.google.android.gms.wallet.PaymentDataRequest

class HyperpaySdkPlugin : FlutterPlugin, MethodCallHandler, ActivityAware, ITransactionListener {

    private lateinit var channel: MethodChannel
    private var context: Context? = null
    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null

    private var paymentProvider: OppPaymentProvider? = null
    private var providerMode: Connect.ProviderMode = Connect.ProviderMode.TEST

    private var pendingResult: Result? = null
    private var checkoutLauncher: ActivityResultLauncher<CheckoutSettings>? = null

    companion object {
        private const val CHANNEL_NAME = "com.hyperpay.sdk/channel"
    }

    // region FlutterPlugin

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
    }

    // endregion

    // region ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        registerCheckoutLauncher()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        registerCheckoutLauncher()
    }

    override fun onDetachedFromActivity() {
        activity = null
        activityBinding = null
        checkoutLauncher = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        activityBinding = null
        checkoutLauncher = null
    }

    // endregion

    // region CheckoutActivityResultContract

    private fun registerCheckoutLauncher() {
        val componentActivity = activity as? ComponentActivity ?: return
        checkoutLauncher = componentActivity.registerForActivityResult(
            CheckoutActivityResultContract()
        ) { result: CheckoutActivityResult ->
            handleCheckoutResult(result)
        }
    }

    private fun handleCheckoutResult(result: CheckoutActivityResult) {
        val pending = pendingResult ?: return
        pendingResult = null

        when {
            result.isCanceled -> {
                pending.success(mapOf(
                    "isSuccess" to false,
                    "isCanceled" to true,
                ))
            }
            result.paymentError != null -> {
                val error = result.paymentError!!
                pending.success(mapOf(
                    "isSuccess" to false,
                    "isCanceled" to false,
                    "errorCode" to error.errorCode,
                    "errorMessage" to error.errorInfo,
                ))
            }
            result.resourcePath != null -> {
                pending.success(mapOf(
                    "isSuccess" to true,
                    "isCanceled" to false,
                    "resourcePath" to result.resourcePath,
                    "transactionType" to "sync",
                ))
            }
            else -> {
                pending.success(mapOf(
                    "isSuccess" to false,
                    "isCanceled" to false,
                    "errorCode" to "UNKNOWN",
                    "errorMessage" to "Unknown checkout result",
                ))
            }
        }
    }

    // endregion

    // region MethodCallHandler

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "setup" -> handleSetup(call, result)
            "checkoutReadyUI" -> handleCheckoutReadyUI(call, result)
            "payCustomUI" -> handlePayCustomUI(call, result)
            "payApplePay" -> handlePayApplePay(result)
            "getPaymentStatus" -> handleGetPaymentStatus(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleSetup(call: MethodCall, result: Result) {
        val mode = call.argument<String>("mode")
        providerMode = if (mode == "live") {
            Connect.ProviderMode.LIVE
        } else {
            Connect.ProviderMode.TEST
        }
        paymentProvider = OppPaymentProvider(context!!, providerMode)
        paymentProvider?.setThreeDSWorkflowListener(object : ThreeDSWorkflowListener {
            override fun onThreeDSChallengeRequired(): Activity? {
                return activity
            }
        })
        result.success(null)
    }

    private fun handleCheckoutReadyUI(call: MethodCall, result: Result) {
        if (activity == null) {
            result.error("NO_ACTIVITY", "Plugin is not attached to an Activity", null)
            return
        }
        if (checkoutLauncher == null) {
            result.error("NO_LAUNCHER", "Checkout launcher not registered. Ensure your Activity extends ComponentActivity.", null)
            return
        }

        val checkoutId = call.argument<String>("checkoutId") ?: run {
            result.error("INVALID_ARGS", "checkoutId is required", null)
            return
        }
        val brands = call.argument<List<String>>("brands") ?: run {
            result.error("INVALID_ARGS", "brands is required", null)
            return
        }
        val shopperResultUrl = call.argument<String>("shopperResultUrl")
        val googlePayConfig = call.argument<Map<String, Any>>("googlePayConfig")
        val lang = call.argument<String>("lang")

        try {
            val paymentBrands = HashSet(brands)
            val checkoutSettings = CheckoutSettings(checkoutId, paymentBrands, providerMode)
            checkoutSettings.setSkipCVVMode(CheckoutSkipCVVMode.FOR_STORED_CARDS)

            if (shopperResultUrl != null) {
                checkoutSettings.shopperResultUrl = "${shopperResultUrl}://callback"
            }

            if (lang != null) {
                checkoutSettings.locale = lang
            }

            // Configure Google Pay if provided
            if (googlePayConfig != null) {
                val merchantId = googlePayConfig["gatewayMerchantId"] as? String ?: ""
                val merchantName = googlePayConfig["merchantName"] as? String ?: ""
                val countryCode = googlePayConfig["countryCode"] as? String ?: ""
                val totalPrice = (googlePayConfig["totalPrice"] as? Number)?.toDouble() ?: 0.0
                val currencyCode = googlePayConfig["currencyCode"] as? String ?: ""

                val googlePayJsonBuilder = com.oppwa.mobile.connect.checkout.googlepay.PaymentDataRequestJsonBuilder()
                    .setGatewayMerchantId(merchantId)
                    .setMerchantName(merchantName)
                    .setTransactionInfo(
                        com.oppwa.mobile.connect.checkout.googlepay.TransactionInfoJsonBuilder()
                            .setTotalPrice(totalPrice)
                            .setTotalPriceStatus("FINAL")
                            .setCountryCode(countryCode)
                            .setCurrencyCode(currencyCode)
                    )
                    .setCardPaymentMethod(
                        com.oppwa.mobile.connect.checkout.googlepay.CardPaymentMethodJsonBuilder()
                    )

                checkoutSettings.setGooglePayPaymentDataRequestJson(googlePayJsonBuilder)
            }

            pendingResult = result
            checkoutLauncher!!.launch(checkoutSettings)
        } catch (e: Exception) {
            result.error("CHECKOUT_ERROR", e.message, null)
        }
    }

    private fun handlePayCustomUI(call: MethodCall, result: Result) {
        if (paymentProvider == null) {
            result.error("NOT_INITIALIZED", "Call setup() before making payments", null)
            return
        }

        val checkoutId = call.argument<String>("checkoutId") ?: run {
            result.error("INVALID_ARGS", "checkoutId is required", null)
            return
        }
        val brand = call.argument<String>("brand") ?: run {
            result.error("INVALID_ARGS", "brand is required", null)
            return
        }
        val cardNumber = call.argument<String>("cardNumber") ?: run {
            result.error("INVALID_ARGS", "cardNumber is required", null)
            return
        }
        val holder = call.argument<String>("holder") ?: run {
            result.error("INVALID_ARGS", "holder is required", null)
            return
        }
        val expiryMonth = call.argument<String>("expiryMonth") ?: run {
            result.error("INVALID_ARGS", "expiryMonth is required", null)
            return
        }
        val expiryYear = call.argument<String>("expiryYear") ?: run {
            result.error("INVALID_ARGS", "expiryYear is required", null)
            return
        }
        val cvv = call.argument<String>("cvv") ?: run {
            result.error("INVALID_ARGS", "cvv is required", null)
            return
        }
        val tokenize = call.argument<Boolean>("tokenize") ?: false
        val shopperResultUrl = call.argument<String>("shopperResultUrl")

        try {
            val paymentParams = CardPaymentParams(
                checkoutId,
                brand,
                cardNumber,
                holder,
                expiryMonth,
                expiryYear,
                cvv
            )

            if (tokenize) {
                paymentParams.tokenizationEnabled = true
            }

            if (shopperResultUrl != null) {
                paymentParams.shopperResultUrl = "${shopperResultUrl}://callback"
            }

            paymentProvider!!.setTransactionListener(this)
            pendingResult = result

            val transaction = Transaction(paymentParams)
            paymentProvider!!.submitTransaction(transaction)
        } catch (e: PaymentException) {
            result.error(
                e.error?.errorCode ?: "PAYMENT_ERROR",
                e.error?.errorInfo ?: e.message ?: "Unknown payment error",
                null
            )
        } catch (e: Exception) {
            result.error("PAYMENT_ERROR", e.message, null)
        }
    }

    private fun handlePayApplePay(result: Result) {
        result.error("UNSUPPORTED", "Apple Pay is not available on Android", null)
    }

    private fun handleGetPaymentStatus(call: MethodCall, result: Result) {
        val resourcePath = call.argument<String>("resourcePath")

        // Payment status verification should be done server-side.
        // This method provides the resource path to the Dart layer so
        // the app can call its backend to verify.
        result.success(mapOf(
            "status" to "PENDING_SERVER_VERIFICATION",
            "resourcePath" to resourcePath,
        ))
    }

    // endregion

    // region ITransactionListener

    override fun transactionCompleted(transaction: Transaction) {
        val pending = pendingResult ?: return
        pendingResult = null

        val resourcePath = transaction.transactionParams?.get("resourcePath") as? String

        if (transaction.transactionType == TransactionType.SYNC) {
            pending.success(mapOf(
                "isSuccess" to true,
                "isCanceled" to false,
                "resourcePath" to resourcePath,
                "transactionType" to "sync",
            ))
        } else {
            // Async transaction — need to redirect
            val redirectUrl = transaction.redirectUrl
            if (redirectUrl != null && activity != null) {
                try {
                    val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(redirectUrl))
                    activity!!.startActivity(browserIntent)
                } catch (_: Exception) {
                    // Browser open failed, still return the result
                }
            }
            pending.success(mapOf(
                "isSuccess" to true,
                "isCanceled" to false,
                "resourcePath" to resourcePath,
                "transactionType" to "async",
            ))
        }
    }

    override fun transactionFailed(transaction: Transaction, paymentError: PaymentError) {
        val pending = pendingResult ?: return
        pendingResult = null

        pending.success(mapOf(
            "isSuccess" to false,
            "isCanceled" to false,
            "errorCode" to paymentError.errorCode,
            "errorMessage" to paymentError.errorInfo,
        ))
    }

    // endregion
}

File: ios/Classes/HyperpaySdkPlugin.swift
import Flutter
import UIKit
import SafariServices
import PassKit
import OPPWAMobile

public class HyperpaySdkPlugin: NSObject, FlutterPlugin {

    private var channel: FlutterMethodChannel?
    private var paymentProvider: OPPPaymentProvider?
    private var checkoutProvider: OPPCheckoutProvider?
    private var pendingResult: FlutterResult?
    private var pendingResourcePath: String?
    private var safariVC: SFSafariViewController?
    private var shopperResultUrl: String?

    // MARK: - FlutterPlugin Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.hyperpay.sdk/channel",
            binaryMessenger: registrar.messenger()
        )
        let instance = HyperpaySdkPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }

    // MARK: - MethodCallHandler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setup":
            handleSetup(call, result: result)
        case "checkoutReadyUI":
            handleCheckoutReadyUI(call, result: result)
        case "payCustomUI":
            handlePayCustomUI(call, result: result)
        case "payApplePay":
            handlePayApplePay(call, result: result)
        case "getPaymentStatus":
            handleGetPaymentStatus(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Setup

    private func handleSetup(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let mode = args["mode"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "mode is required", details: nil))
            return
        }

        let providerMode: OPPProviderMode = mode == "live" ? .live : .test
        paymentProvider = OPPPaymentProvider(mode: providerMode)
        result(nil)
    }

    // MARK: - ReadyUI Checkout

    private func handleCheckoutReadyUI(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let checkoutId = args["checkoutId"] as? String,
              let brands = args["brands"] as? [String] else {
            result(FlutterError(code: "INVALID_ARGS", message: "checkoutId and brands are required", details: nil))
            return
        }

        guard let provider = paymentProvider else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Call setup() before making payments", details: nil))
            return
        }

        let shopperUrl = args["shopperResultUrl"] as? String
        let applePayConfig = args["applePayConfig"] as? [String: Any]

        // Configure checkout settings
        let checkoutSettings = OPPCheckoutSettings()
        checkoutSettings.paymentBrands = brands
        checkoutSettings.storePaymentDetails = .prompt

        if let shopperUrl = shopperUrl {
            checkoutSettings.shopperResultURL = "\(shopperUrl)://callback"
            self.shopperResultUrl = shopperUrl
        }

        // Apple Pay configuration
        if let applePayConfig = applePayConfig {
            let merchantId = applePayConfig["merchantId"] as? String ?? ""
            let countryCode = applePayConfig["countryCode"] as? String ?? ""
            let currencyCode = applePayConfig["currencyCode"] as? String ?? ""
            let amount = applePayConfig["amount"] as? Double ?? 0.0
            let companyName = applePayConfig["companyName"] as? String ?? ""

            let paymentRequest = OPPPaymentProvider.paymentRequest(
                withMerchantIdentifier: merchantId,
                countryCode: countryCode
            )
            paymentRequest.currencyCode = currencyCode
            paymentRequest.paymentSummaryItems = [
                PKPaymentSummaryItem(label: companyName, amount: NSDecimalNumber(value: amount))
            ]
            checkoutSettings.applePayPaymentRequest = paymentRequest
        }

        // Language
        if let lang = args["lang"] as? String {
            checkoutSettings.language = lang
        }

        // Create checkout provider
        checkoutProvider = OPPCheckoutProvider(
            paymentProvider: provider,
            checkoutID: checkoutId,
            settings: checkoutSettings
        )
        checkoutProvider?.delegate = self

        pendingResult = result

        // Present checkout
        DispatchQueue.main.async { [weak self] in
            self?.checkoutProvider?.presentCheckout(
                forSubmittingTransactionCompletionHandler: { [weak self] (transaction, error) in
                    self?.handleCheckoutCompletion(transaction: transaction, error: error)
                },
                cancelHandler: { [weak self] in
                    self?.handleCheckoutCancellation()
                }
            )
        }
    }

    private func handleCheckoutCompletion(transaction: OPPTransaction?, error: Error?) {
        guard let pending = pendingResult else { return }

        if let error = error as NSError? {
            let detail = "domain=\(error.domain), code=\(error.code), userInfo=\(error.userInfo)"
            pending([
                "isSuccess": false,
                "isCanceled": false,
                "errorCode": "\(error.code)",
                "errorMessage": "\(error.localizedDescription) [\(detail)]",
            ] as [String: Any])
            pendingResult = nil
            return
        }

        guard let transaction = transaction else {
            pending([
                "isSuccess": false,
                "isCanceled": false,
                "errorCode": "UNKNOWN",
                "errorMessage": "No transaction returned",
            ] as [String: Any])
            pendingResult = nil
            return
        }

        if transaction.type == .synchronous {
            pending([
                "isSuccess": true,
                "isCanceled": false,
                "resourcePath": transaction.resourcePath ?? "",
                "transactionType": "sync",
            ] as [String: Any])
            pendingResult = nil
        } else {
            // Async — keep pendingResult alive, wait for URL scheme callback
            // OPPCheckoutProvider handles the redirect UI internally
            pendingResourcePath = transaction.resourcePath
        }
    }

    private func handleCheckoutCancellation() {
        guard let pending = pendingResult else { return }
        pending([
            "isSuccess": false,
            "isCanceled": true,
        ] as [String: Any])
        pendingResult = nil
    }

    // MARK: - CustomUI Payment

    private func handlePayCustomUI(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let checkoutId = args["checkoutId"] as? String,
              let brand = args["brand"] as? String,
              let cardNumber = args["cardNumber"] as? String,
              let holder = args["holder"] as? String,
              let expiryMonth = args["expiryMonth"] as? String,
              let expiryYear = args["expiryYear"] as? String,
              let cvv = args["cvv"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "All card fields are required", details: nil))
            return
        }

        guard let provider = paymentProvider else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Call setup() before making payments", details: nil))
            return
        }

        let tokenize = args["tokenize"] as? Bool ?? false
        let shopperUrl = args["shopperResultUrl"] as? String

        NSLog("[HyperPay] CustomUI native params: checkoutId=%@, brand=%@, card=%@, holder=%@, expiry=%@/%@, cvv=%@, tokenize=%d, shopperUrl=%@",
              checkoutId, brand, cardNumber, holder, expiryMonth, expiryYear, cvv, tokenize ? 1 : 0, shopperUrl ?? "nil")

        do {
            let paymentParams = try OPPCardPaymentParams(
                checkoutID: checkoutId,
                paymentBrand: brand,
                holder: holder,
                number: cardNumber,
                expiryMonth: expiryMonth,
                expiryYear: expiryYear,
                cvv: cvv
            )
            NSLog("[HyperPay] OPPCardPaymentParams created successfully")

            if tokenize {
                paymentParams.isTokenizationEnabled = true
            }

            if let shopperUrl = shopperUrl {
                paymentParams.shopperResultURL = "\(shopperUrl)://callback"
                self.shopperResultUrl = shopperUrl
            }

            let transaction = OPPTransaction(paymentParams: paymentParams)

            pendingResult = result

            NSLog("[HyperPay] Submitting transaction to provider...")

            provider.submitTransaction(transaction) { [weak self] (transaction, error) in
                guard let self = self, let pending = self.pendingResult else { return }

                if let error = error as NSError? {
                    self.pendingResult = nil
                    let detail = "domain=\(error.domain), code=\(error.code), userInfo=\(error.userInfo)"
                    pending([
                        "isSuccess": false,
                        "isCanceled": false,
                        "errorCode": "\(error.code)",
                        "errorMessage": "\(error.localizedDescription) [\(detail)]",
                    ] as [String: Any])
                    return
                }

                if transaction.type == .synchronous {
                    self.pendingResult = nil
                    pending([
                        "isSuccess": true,
                        "isCanceled": false,
                        "resourcePath": transaction.resourcePath ?? "",
                        "transactionType": "sync",
                    ] as [String: Any])
                } else {
                    // Async — keep pendingResult alive, wait for URL scheme callback
                    self.pendingResourcePath = transaction.resourcePath
                    if let redirectUrl = transaction.redirectURL {
                        DispatchQueue.main.async {
                            self.openSafariVC(url: redirectUrl)
                        }
                    }
                }
            }
        } catch {
            NSLog("[HyperPay] OPPCardPaymentParams creation failed: %@", error.localizedDescription)
            result(FlutterError(
                code: "PAYMENT_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }

    // MARK: - Apple Pay

    private func handlePayApplePay(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let checkoutId = args["checkoutId"] as? String,
              let merchantId = args["merchantId"] as? String,
              let countryCode = args["countryCode"] as? String,
              let currencyCode = args["currencyCode"] as? String,
              let amount = args["amount"] as? Double,
              let companyName = args["companyName"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "All Apple Pay fields are required", details: nil))
            return
        }

        guard let provider = paymentProvider else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Call setup() before making payments", details: nil))
            return
        }

        let paymentRequest = OPPPaymentProvider.paymentRequest(
            withMerchantIdentifier: merchantId,
            countryCode: countryCode
        )
        paymentRequest.currencyCode = currencyCode
        paymentRequest.paymentSummaryItems = [
            PKPaymentSummaryItem(label: companyName, amount: NSDecimalNumber(value: amount))
        ]

        do {
            let params = try OPPApplePayPaymentParams(
                checkoutID: checkoutId,
                paymentBrand: "APPLEPAY"
            )

            pendingResult = result

            // Present Apple Pay
            if let rootVC = self.rootViewController() {
                let applePayController = PKPaymentAuthorizationViewController(paymentRequest: paymentRequest)
                applePayController?.delegate = self

                rootVC.present(applePayController!, animated: true) {
                    // Store params for submission after Apple Pay authorization
                    self.pendingApplePayParams = params
                    self.pendingApplePayProvider = provider
                }
            } else {
                result(FlutterError(code: "NO_VIEW_CONTROLLER", message: "Cannot present Apple Pay", details: nil))
            }
        } catch {
            result(FlutterError(
                code: "APPLE_PAY_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }

    private var pendingApplePayParams: OPPApplePayPaymentParams?
    private var pendingApplePayProvider: OPPPaymentProvider?

    // MARK: - Payment Status

    private func handleGetPaymentStatus(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let resourcePath = args?["resourcePath"] as? String

        // Payment status verification should be done server-side.
        result([
            "status": "PENDING_SERVER_VERIFICATION",
            "resourcePath": resourcePath as Any,
        ] as [String: Any])
    }

    // MARK: - Safari VC for Async Payments

    private func openSafariVC(url: URL) {
        guard let rootVC = rootViewController() else { return }

        safariVC = SFSafariViewController(url: url)
        safariVC?.delegate = self
        rootVC.present(safariVC!, animated: true, completion: nil)
    }

    // MARK: - URL Scheme Handling

    public func application(
        _ app: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        guard let shopperUrl = shopperResultUrl,
              url.scheme?.caseInsensitiveCompare(shopperUrl) == .orderedSame else {
            return false
        }

        // Dismiss Safari VC (CustomUI async)
        if let safariVC = safariVC {
            safariVC.dismiss(animated: true, completion: nil)
            self.safariVC = nil
        }

        // Dismiss ReadyUI checkout and resolve pending result
        if let pending = pendingResult {
            let resourcePath = pendingResourcePath ?? ""
            pendingResourcePath = nil

            if let checkoutProvider = checkoutProvider {
                checkoutProvider.dismissCheckout(animated: true) {
                    pending([
                        "isSuccess": true,
                        "isCanceled": false,
                        "resourcePath": resourcePath,
                        "transactionType": "async",
                    ] as [String: Any])
                }
            } else {
                pending([
                    "isSuccess": true,
                    "isCanceled": false,
                    "resourcePath": resourcePath,
                    "transactionType": "async",
                ] as [String: Any])
            }
            pendingResult = nil
        }

        return true
    }

    // MARK: - Helpers

    private func rootViewController() -> UIViewController? {
        if #available(iOS 15.0, *) {
            let scenes = UIApplication.shared.connectedScenes
            let windowScene = scenes.first as? UIWindowScene
            return windowScene?.windows.first(where: { $0.isKeyWindow })?.rootViewController
        } else {
            return UIApplication.shared.windows.first(where: { $0.isKeyWindow })?.rootViewController
        }
    }
}

// MARK: - OPPCheckoutProviderDelegate

extension HyperpaySdkPlugin: OPPCheckoutProviderDelegate {
    public func checkoutProvider(
        _ checkoutProvider: OPPCheckoutProvider,
        continueSubmitting transaction: OPPTransaction,
        completion: @escaping (String, Bool) -> Void
    ) {
        // Continue with the transaction by default
        completion(transaction.resourcePath ?? "", true)
    }
}

// MARK: - SFSafariViewControllerDelegate

extension HyperpaySdkPlugin: SFSafariViewControllerDelegate {
    public func safariViewControllerDidFinish(_ controller: SFSafariViewController) {
        safariVC = nil
    }
}

// MARK: - PKPaymentAuthorizationViewControllerDelegate

extension HyperpaySdkPlugin: PKPaymentAuthorizationViewControllerDelegate {
    public func paymentAuthorizationViewControllerDidFinish(
        _ controller: PKPaymentAuthorizationViewController
    ) {
        controller.dismiss(animated: true) { [weak self] in
            guard let self = self else { return }
            if self.pendingResult != nil && self.pendingApplePayParams == nil {
                // User cancelled without authorizing
                let pending = self.pendingResult
                self.pendingResult = nil
                pending?([
                    "isSuccess": false,
                    "isCanceled": true,
                ] as [String: Any])
            }
        }
    }

    public func paymentAuthorizationViewController(
        _ controller: PKPaymentAuthorizationViewController,
        didAuthorizePayment payment: PKPayment,
        handler completion: @escaping (PKPaymentAuthorizationResult) -> Void
    ) {
        guard let params = pendingApplePayParams,
              let provider = pendingApplePayProvider else {
            completion(PKPaymentAuthorizationResult(status: .failure, errors: nil))
            return
        }

        let paramsWithToken = try? OPPApplePayPaymentParams(
            checkoutID: params.checkoutID,
            paymentBrand: "APPLEPAY",
            tokenData: payment.token.paymentData
        )
        guard let finalParams = paramsWithToken else {
            completion(PKPaymentAuthorizationResult(status: .failure, errors: nil))
            return
        }

        let transaction = OPPTransaction(paymentParams: finalParams)
        provider.submitTransaction(transaction) { [weak self] (transaction, error) in
            guard let self = self else { return }
            let pending = self.pendingResult
            self.pendingResult = nil
            self.pendingApplePayParams = nil
            self.pendingApplePayProvider = nil

            if let error = error {
                completion(PKPaymentAuthorizationResult(status: .failure, errors: nil))
                pending?([
                    "isSuccess": false,
                    "isCanceled": false,
                    "errorCode": "\((error as NSError).code)",
                    "errorMessage": error.localizedDescription,
                ] as [String: Any])
                return
            }

            completion(PKPaymentAuthorizationResult(status: .success, errors: nil))
            pending?([
                "isSuccess": true,
                "isCanceled": false,
                "resourcePath": transaction.resourcePath ?? "",
                "transactionType": "sync",
            ] as [String: Any])
        }
    }
}
